lib/route/nh.c

```diff
@@ -0,0 +1,529 @@
+#include <netlink/route/nh.h>
+#include <netlink/hashtable.h>
+
+/** @cond SKIP */
+#define NH_ATTR_FLAGS			(1 << 0)
+#define NH_ATTR_ID			(1 << 1)
+#define NH_ATTR_GATEWAY			(1 << 2)
+#define NH_ATTR_OIF			(1 << 3)
+#define NH_ATTR_GROUP			(1 << 4)
+#define NH_ATTR_MASTER			(1 << 5)
+#define NH_ATTR_FLAG_FDB		(1 << 6)
+#define NH_ATTR_FLAG_BLACKHOLE		(1 << 7)
+#define NH_ATTR_FLAG_GROUPS		(1 << 8)
+/** @endcond */
+
+struct nla_policy rtnl_nh_policy[NHA_MAX+1] = {
+	[NHA_UNSPEC]		= { .type = NLA_UNSPEC },
+	[NHA_ID]		= { .type = NLA_U32 },
+	[NHA_GROUP]		= { .type = NLA_NESTED },
+	[NHA_GROUP_TYPE] 	= { .type = NLA_U16 },
+	[NHA_BLACKHOLE]  	= { .type = NLA_UNSPEC },
+	[NHA_OIF]		= { .type = NLA_U32 },
+	[NHA_MASTER]		= { .type = NLA_U32 },
+};
+
+static struct nl_cache_ops rtnl_nh_ops;
+static struct nl_object_ops nh_obj_ops;
+
+static nl_nh_group_t *rtnl_nh_grp_alloc(int len)
+{
+	nl_nh_group_t *nhg;
+
+	if (!(nhg = calloc(1, sizeof(*nhg))))
+		return NULL;
+
+	if (!(nhg->entries = calloc(len, sizeof(*nhg->entries))))
+	{
+		free (nhg);
+		return NULL;
+	}
+
+	nhg->ce_refcnt = 1;
+
+	NL_DBG(4, "Allocated new NHG object %p\n", nhg);
+
+	return nhg;
+}
+
+static void rtnl_nh_grp_free(nl_nh_group_t *nhg)
+{
+	if (!nhg)
+		return;
+
+	if (nhg->ce_refcnt > 0)
+		NL_DBG(1, "Warning: Freeing NHG object in use...\n");
+
+	NL_DBG(4, "Freed NHG object %p\n", nhg);
+	free(nhg);
+
+	return;
+}
+
+static void rtnl_nh_grp_put(nl_nh_group_t *nhg)
+{
+	if (!nhg)
+		return;
+
+	nhg->ce_refcnt--;
+	NL_DBG(4, "Returned NHG object reference %p, %i remaining\n",
+				 nhg, nhg->ce_refcnt);
+
+	if (nhg->ce_refcnt < 0)
+		BUG();
+
+	if (nhg->ce_refcnt <= 0)
+		rtnl_nh_grp_free(nhg);
+
+	return;
+}
+
+struct rtnl_nh *rtnl_nh_alloc(void)
+{
+	return (struct rtnl_nh *) nl_object_alloc(&nh_obj_ops);
+}
+
+static int nh_clone(struct nl_object *_src, struct nl_object *_dst)
+{
+	struct rtnl_nh *dst = nl_object_priv(_dst);
+	struct rtnl_nh *src = nl_object_priv(_src);
+
+	dst->nh_flags = src->nh_flags;
+	dst->nh_family = src->nh_family;
+	dst->nh_id = src->nh_id;
+	dst->nh_oif = src->nh_oif;
+	dst->nh_master = src->nh_master;
+	dst->ce_mask = src->ce_mask;
+
+	return 0;
+}
+
+static void nh_free(struct nl_object *obj)
+{
+	struct rtnl_nh *nh = nl_object_priv (obj);
+	nl_addr_put(nh->nh_gateway);
+	if (nh->nh_group)
+		rtnl_nh_grp_put(nh->nh_group);
+}
+
+void rtnl_nh_put(struct rtnl_nh *nh)
+{
+	struct nl_object *obj = (struct nl_object*)nh;
+	nl_object_put (obj);
+}
+
+static void nexthop_keygen(struct nl_object *obj, uint32_t *hashkey,
+				uint32_t table_sz)
+{
+	struct rtnl_nh *nexthop = nl_object_priv (obj);
+	unsigned int nhkey_sz;
+	struct nexthop_hash_key {
+		uint32_t	nh_id;
+	} __attribute__((packed)) nhkey;
+
+	nhkey_sz = sizeof(nhkey);
+	nhkey.nh_id = nexthop->nh_id;
+
+	*hashkey = nl_hash(&nhkey, nhkey_sz, 0) % table_sz;
+
+	NL_DBG(5, "nexthop %p key (nhid %d) keysz %d, hash 0x%x\n",
+				 nexthop, nhkey.nh_id, nhkey_sz, *hashkey);
```

Comment:
I think a lot of the debug logging (all?) should be dropped. It doesn't seem useful when a library logs to stdout/stderr.

Reply:
done